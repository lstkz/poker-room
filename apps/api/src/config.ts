import dotenv from 'dotenv';
import Path from 'path';

dotenv.config({
  path: Path.join(__dirname, '../../../.env'),
});

export const PORT = process.env.PORT || 3000;

if (!process.env.MONGO_URL) {
  throw new Error('MONGO_URL is not set');
}

export const MONGO_URL = process.env.MONGO_URL;

if (!process.env.MONGO_DB_NAME) {
  throw new Error('MONGO_DB_NAME is not set');
}

export const MONGO_DB_NAME = process.env.MONGO_DB_NAME;

export const LOG_LEVEL: 'debug' | 'info' =
  process.env.NODE_ENV === 'development' ? 'debug' : 'info';
