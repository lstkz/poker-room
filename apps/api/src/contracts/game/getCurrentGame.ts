import * as R from 'remeda';
import { ObjectID } from 'mongodb';
import { S } from 'schema';
import { Game } from 'shared';
import { GameCollection } from '../../collections/Game';
import { TableCollection } from '../../collections/Table';
import { TableNotFoundError } from '../../common/errors';
import { createContract, createRpcBinding } from '../../lib';
import { AppUser } from '../../types';
import { UserCollection, UserModel } from '../../collections/User';

export const getCurrentGame = createContract('game.getCurrentGame')
  .params('user', 'tableId')
  .schema({
    user: S.object().as<AppUser>(),
    tableId: S.string().objectId(),
  })
  .returns<Game>()
  .fn(async (user, tableId) => {
    const table = await TableCollection.findOne({
      _id: ObjectID.createFromHexString(tableId),
    });
    if (!table) {
      throw new TableNotFoundError();
    }
    const playerIds = table.players.map(x => x.userId);
    const players: UserModel[] = playerIds.length
      ? await UserCollection.findAll({
          _id: {
            $in: playerIds,
          },
        })
      : [];
    const playerMap = R.indexBy(players, x => x._id);
    const game = await GameCollection.findOneOrThrow({ _id: table.gameId });
    return {
      id: game._id.toHexString(),
      tableId: game.tableId.toHexString(),
      ...R.pick(game, [
        'isStarted',
        'isDone',
        'dealerPosition',
        'stakes',
        'pot',
        'betMap',
        'currentBets',
      ]),
      phases: game.phases.map(item => ({
        cards: item.cards,
        moves: item.moves.map(move => ({
          userId: move.userId.toHexString(),
          moveType: move.moveType,
          amount: move.amount,
        })),
        type: item.type,
      })),
      players: game.players.map(item => ({
        user: {
          id: item.userId.toHexString(),
          username: playerMap[item.userId.toHexString()].username,
        },
        hand: item.userId.equals(user.id) ? item.hand : null,
        money: item.money,
        seat: item.seat,
      })),
    };
  });

export const getCurrentGameRpc = createRpcBinding({
  injectUser: true,
  signature: 'game.getCurrentGame',
  handler: getCurrentGame,
});
