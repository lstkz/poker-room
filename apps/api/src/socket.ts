import http from 'http';
import SocketIO from 'socket.io';
import { AccessTokenCollection } from './collections/AccessToken';

let io: SocketIO.Server | null = null;

export const userSocketMap: Record<string, Array<SocketIO.Socket>> = {};
export const userTableMap: Record<string, Array<SocketIO.Socket>> = {};
export const socketUserIdMap: Record<string, string> = {};

export function getIO() {
  if (!io) {
    throw new Error('IO not set');
  }
  return io;
}

function _removeArrayItem<T>(items: T[], item: T) {
  if (!items) {
    return;
  }
  const idx = items.indexOf(item);
  if (idx !== -1) {
    items.splice(items.indexOf(item), 1);
  }
}

export function initSocket(server: http.Server) {
  io = SocketIO(server);

  io.use(async (socket, next) => {
    try {
      const tokenEntity = await AccessTokenCollection.findOne({
        _id: socket.handshake.query.accessToken,
      });
      if (!tokenEntity) {
        return next(new Error('authentication error'));
      }
      socketUserIdMap[socket.id] = tokenEntity.userId;
      next();
    } catch (e) {
      return next(e);
    }
  });

  io.on('connect', socket => {
    const userId = socketUserIdMap[socket.id];
    if (!userSocketMap[userId]) {
      userSocketMap[userId] = [];
    }
    const sockets = userSocketMap[userId];
    sockets.push(socket);
    socket.on('disconnect', () => {
      _removeArrayItem(sockets, socket);
      Object.values(userTableMap).forEach(arr => {
        _removeArrayItem(arr, socket);
      });
    });

    socket.on('join-table', data => {
      if (!userTableMap[data.id]) {
        userTableMap[data.id] = [];
      }
      userTableMap[data.id].push(socket);
    });
    socket.on('leave-table', data => {
      _removeArrayItem(userTableMap[data.id], socket);
    });
  });
}
