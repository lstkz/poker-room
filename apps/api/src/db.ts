import * as R from 'remeda';
import Path from 'path';
import fs from 'fs';
import { MONGO_URL, MONGO_DB_NAME } from './config';
import {
  MongoClient,
  CollectionAggregationOptions,
  AggregationCursor,
  MongoCallback,
  FilterQuery,
  MongoCountPreferences,
  FindOneOptions,
  FindOneAndDeleteOption,
  FindOneAndUpdateOption,
  UpdateQuery,
  FindOneAndReplaceOption,
  CollectionInsertOneOptions,
  InsertOneWriteOpResult,
  Cursor,
  FindAndModifyWriteOpResultObject,
  IndexSpecification,
  DeleteWriteOpResultObject,
  CommonOptions,
  InsertWriteOpResult,
  CollectionInsertManyOptions,
  OptionalId,
  WithId,
} from 'mongodb';

let client: MongoClient | null = null;

export async function disconnect() {
  client?.close();
}

export async function connect() {
  if (!client || !client.isConnected()) {
    client = new MongoClient(MONGO_URL, {
      useNewUrlParser: true,
      useUnifiedTopology: true,
    });
  }
  await client.connect();

  return client;
}

function getClient() {
  if (!client) {
    throw new Error('client is not set');
  }
  return client;
}

interface DbCollection<TSchema> {
  aggregate<T>(
    pipeline?: object[],
    options?: CollectionAggregationOptions,
    callback?: MongoCallback<AggregationCursor<T>>
  ): AggregationCursor<T>;
  insertOne(
    docs: OptionalId<TSchema>,
    options?: CollectionInsertOneOptions
  ): Promise<InsertOneWriteOpResult<WithId<TSchema>>>;
  insertMany(
    docs: Array<OptionalId<TSchema>>,
    options?: CollectionInsertManyOptions
  ): Promise<InsertWriteOpResult<WithId<TSchema>>>;
  countDocuments(
    query?: FilterQuery<TSchema>,
    options?: MongoCountPreferences
  ): Promise<number>;
  find<T = TSchema>(
    query: FilterQuery<TSchema>,
    options?: FindOneOptions<T extends TSchema ? TSchema : T>
  ): Cursor<T>;
  findOne<T = TSchema>(
    filter: FilterQuery<TSchema>,
    options?: FindOneOptions<T extends TSchema ? TSchema : T>
  ): Promise<T | null>;
  findOneOrThrow<T = TSchema>(
    filter: FilterQuery<TSchema>,
    options?: FindOneOptions<T extends TSchema ? TSchema : T>
  ): Promise<T>;
  findOneAndDelete(
    filter: FilterQuery<TSchema>,
    options?: FindOneAndDeleteOption<TSchema>
  ): Promise<FindAndModifyWriteOpResultObject<TSchema>>;
  findOneAndReplace(
    filter: FilterQuery<TSchema>,
    replacement: object,
    options?: FindOneAndReplaceOption<TSchema>
  ): Promise<FindAndModifyWriteOpResultObject<TSchema>>;
  findOneAndUpdate(
    filter: FilterQuery<TSchema>,
    update: UpdateQuery<TSchema> | TSchema,
    options?: FindOneAndUpdateOption<TSchema>
  ): Promise<FindAndModifyWriteOpResultObject<TSchema>>;
  deleteMany(
    filter: FilterQuery<TSchema>,
    options?: CommonOptions
  ): Promise<DeleteWriteOpResultObject>;
  deleteOne(
    filter: FilterQuery<TSchema>,
    options?: CommonOptions
  ): Promise<DeleteWriteOpResultObject>;
  update(model: TSchema, fields: Array<keyof TSchema>): Promise<void>;
}

export function createCollection<T>(
  collectionName: string,
  indexes?: IndexSpecification[]
): DbCollection<T> {
  const _getCollection = () => {
    const client = getClient();
    const db = client.db(
      process.env.JEST_WORKER_ID
        ? `jest-${process.env.JEST_WORKER_ID}`
        : MONGO_DB_NAME
    );
    return db.collection<T>(collectionName);
  };

  const ret: DbCollection<T> = {
    aggregate(...args) {
      return _getCollection().aggregate(...args);
    },
    insertOne(...args) {
      return _getCollection().insertOne(...args);
    },
    insertMany(...args) {
      return _getCollection().insertMany(...args);
    },
    countDocuments(...args) {
      return _getCollection().countDocuments(...args);
    },
    find(...args) {
      return _getCollection().find(...args);
    },
    findOne(...args) {
      return _getCollection().findOne(...args);
    },
    async findOneOrThrow(...args) {
      const ret = await _getCollection().findOne(...args);
      if (!ret) {
        throw new Error(
          `Entity ${JSON.stringify(args[0])} not found in ${collectionName}`
        );
      }
      return ret;
    },
    findOneAndDelete(...args) {
      return _getCollection().findOneAndDelete(...args);
    },
    findOneAndReplace(...args) {
      return _getCollection().findOneAndReplace(...args);
    },
    findOneAndUpdate(...args) {
      return _getCollection().findOneAndUpdate(...args);
    },
    deleteMany(...args) {
      return _getCollection().deleteMany(...args);
    },
    deleteOne(...args) {
      return _getCollection().deleteOne(...args);
    },
    async update(model: any, fields) {
      if (model._id == null) {
        throw new Error('_id not defined');
      }
      await this.findOneAndUpdate(
        {
          _id: model._id,
        },
        {
          $set: R.pick(model, fields),
        }
      );
    },
  };

  (ret as any).initIndex = () => {
    if (!indexes || indexes.length === 0) {
      return;
    }
    return _getCollection().createIndexes(indexes);
  };

  return ret;
}

export function getAllCollection(): Array<DbCollection<unknown>> {
  return R.pipe(
    fs.readdirSync(Path.join(__dirname, './collections')),
    R.map(name => Object.values(require('./collections/' + name))),
    R.flatten
  ) as any;
}

export function initIndexes() {
  return Promise.all(
    getAllCollection().map((collection: any) => collection.initIndex())
  );
}
