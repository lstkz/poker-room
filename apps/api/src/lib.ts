import { ContractMeta, initialize } from 'contract';
import { StringSchema } from 'schema';
import { LOG_LEVEL } from './config';
import { AppEvent, AppEventType } from './types';

export interface CreateRpcBindingOptions {
  verified?: true;
  injectUser?: boolean;
  public?: true;
  admin?: true;
  signature: string;
  handler: ((...args: any[]) => any) & ContractMeta<any>;
}

export interface BaseBinding<T, U> {
  isBinding: boolean;
  type: T;
  options: U;
}

export interface RpcBinding
  extends BaseBinding<'rpc', CreateRpcBindingOptions> {}

export function createRpcBinding(options: CreateRpcBindingOptions): RpcBinding {
  return {
    isBinding: true,
    type: 'rpc',
    options,
  };
}

type ExtractPayload<T> = T extends { payload: infer S } ? S : never;

type ExtractEvent<T> = AppEvent extends { type: infer K }
  ? K extends T
    ? ExtractPayload<Pick<AppEvent, 'payload'>>
    : never
  : never;

export interface CreateEventBindingOptions<T extends AppEventType> {
  type: T;
  handler: (event: ExtractEvent<T>) => Promise<void>;
}

export interface EventBinding<T extends AppEventType>
  extends BaseBinding<'event', CreateEventBindingOptions<T>> {}

export function createEventBinding<T extends AppEventType>(
  options: CreateEventBindingOptions<T>
): EventBinding<T> {
  return {
    isBinding: true,
    type: 'event',
    options,
  };
}

export const { createContract } = initialize({
  debug: LOG_LEVEL === 'debug',
});

declare module 'schema/src/StringSchema' {
  interface StringSchema {
    objectId(): this;
  }
}

StringSchema.prototype.objectId = function objectId(this: StringSchema) {
  return this.regex(/^[a-f0-9]{24}$/);
};
