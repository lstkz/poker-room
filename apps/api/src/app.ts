import express from 'express';
import Path from 'path';
import http from 'http';
import bodyParser from 'body-parser';
import compression from 'compression';
import { logger } from './common/logger';
import { PORT } from './config';
import { connect, initIndexes } from './db';
import { domainMiddleware } from './middlewares/domainMiddleware';
import { errorHandlerMiddleware } from './middlewares/errorHandlerMiddleware';
import { notFoundHandlerMiddleware } from './middlewares/notFoundHandlerMiddleware';
import loadRoutes from './common/loadRoutes';

connect().then(async () => {
  await initIndexes();
  const app = express();
  app.set('port', PORT);
  app.use(domainMiddleware);
  app.use(compression());
  app.use(bodyParser.json());
  const apiRouter = express.Router();
  loadRoutes(apiRouter);
  app.use('/api', apiRouter);
  app.use(express.static(Path.join(__dirname, '../../front/build')));

  app.use(errorHandlerMiddleware);
  app.use(notFoundHandlerMiddleware);
  const server = http.createServer(app);
  server.listen(app.get('port'), '0.0.0.0', () => {
    logger.info(
      `Express HTTP server listening on port ${app.get('port')} in ${
        process.env.NODE_ENV
      } mode`
    );
  });
});
