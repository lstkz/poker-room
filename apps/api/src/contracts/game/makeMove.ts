import { ObjectID } from 'mongodb';
import { S } from 'schema';
import { MoveType, moveTypes } from 'shared';
import { GameCollection } from '../../collections/Game';
import {
  getActPlayer,
  processMove,
  processNextPhase,
} from '../../common/engine';
import { BadRequestError } from '../../common/errors';
import { withTransaction } from '../../db';
import { dispatch } from '../../events/dispatch';
import { createContract, createRpcBinding } from '../../lib';
import { AppUser } from '../../types';
import { safeKeys } from '../../common/helper';

export const makeMove = createContract('game.makeMove')
  .params('user', 'values')
  .schema({
    user: S.object().as<AppUser>(),
    values: S.object().keys({
      gameId: S.string().objectId(),
      moveType: S.enum().literal<MoveType>(...moveTypes),
      raiseAmount: S.number().optional(),
    }),
  })
  .returns<void>()
  .fn(async (user, values) => {
    await withTransaction(async () => {
      const game = await GameCollection.findOneOrThrow({
        _id: ObjectID.createFromHexString(values.gameId),
      });
      if (game.isDone) {
        throw new BadRequestError('Game is finished');
      }
      const actPlayer = getActPlayer(game);
      if (!actPlayer.userId.equals(user.id)) {
        throw new BadRequestError("It's not your turn");
      }
      processMove(
        {
          moveType: values.moveType,
          raiseAmount: values.raiseAmount,
        },
        game
      );
      await processNextPhase(game);
      await GameCollection.update(game, safeKeys(game));
      dispatch({
        type: 'GAME_UPDATED',
        payload: { gameId: game._id.toHexString() },
      });
    });
  });

export const makeMoveRpc = createRpcBinding({
  injectUser: true,
  signature: 'game.makeMove',
  handler: makeMove,
});
