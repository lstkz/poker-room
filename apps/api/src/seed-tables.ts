import { GameCollection } from './collections/Game';
import { TableCollection } from './collections/Table';
import { createTable } from './contracts/table/createTable';
import { connect, disconnect } from './db';

const tables = [
  { name: 'Saturn', stakes: 50, maxSeats: 6 },
  { name: 'Mars', stakes: 50, maxSeats: 6 },
  { name: 'Alpha', stakes: 100, maxSeats: 6 },
  { name: 'Beta', stakes: 100, maxSeats: 6 },
];

async function start() {
  await connect();
  await GameCollection.deleteMany({});
  await TableCollection.deleteMany({});
  await Promise.all(tables.map(values => createTable(values)));
  await disconnect();
}

start();
