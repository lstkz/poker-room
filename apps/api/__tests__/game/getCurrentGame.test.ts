import { getCurrentGame } from '../../src/contracts/game/getCurrentGame';
import { makeMove } from '../../src/contracts/game/makeMove';
import { disconnect } from '../../src/db';
import {
  execContract,
  generateSampleTables,
  generateSampleUsers,
  getTestGameId,
  getTestTableId,
  initDb,
  joinTableWithDispatch,
  resetDb,
} from '../helper';

jest.mock('../../src/events/dispatch', () => {
  return {
    dispatch() {
      //
    },
  };
});

jest.mock('../../src/common/random', () => {
  return {
    async randomInt() {
      return 0;
    },
  };
});

beforeAll(async () => {
  await initDb();
});

afterAll(disconnect);

beforeEach(async () => {
  await resetDb();
  await generateSampleUsers();
  await generateSampleTables();
});

it('should return a game with no players', async () => {
  const ret = await execContract(
    getCurrentGame,
    {
      tableId: getTestTableId(1),
    },
    'token_1'
  );
  expect(ret).toMatchSnapshot();
});

it('should return a game with 3 players (new games)', async () => {
  await joinTableWithDispatch({ userNr: 1, seat: 1, tableNr: 1 });
  await joinTableWithDispatch({ userNr: 2, seat: 2, tableNr: 1 });
  await joinTableWithDispatch({ userNr: 3, seat: 3, tableNr: 1 });
  const ret = await execContract(
    getCurrentGame,
    {
      tableId: getTestTableId(1),
    },
    'token_1'
  );
  expect(ret).toMatchSnapshot();
});

xit('should return a game with 3 players (with moves)', async () => {
  await joinTableWithDispatch({ userNr: 1, seat: 1, tableNr: 1 });
  await joinTableWithDispatch({ userNr: 2, seat: 2, tableNr: 1 });
  await joinTableWithDispatch({ userNr: 3, seat: 3, tableNr: 1 });
  await execContract(
    makeMove,
    {
      values: {
        gameId: getTestGameId(1),
        moveType: 'call',
      },
    },
    'token_1'
  );
  await execContract(
    makeMove,
    {
      values: {
        gameId: getTestGameId(1),
        moveType: 'call',
      },
    },
    'token_2'
  );
  await execContract(
    makeMove,
    {
      values: {
        gameId: getTestGameId(1),
        moveType: 'check',
      },
    },
    'token_3'
  );
  const ret = await execContract(
    getCurrentGame,
    {
      tableId: getTestTableId(1),
    },
    'token_1'
  );
  expect(ret).toMatchSnapshot();
});
