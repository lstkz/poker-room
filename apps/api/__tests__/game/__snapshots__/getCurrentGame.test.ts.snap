// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return a game with 3 players (new games) 1`] = `
Object {
  "betMap": Object {
    "000000000000000000000002": 0.25,
    "000000000000000000000003": 0.5,
  },
  "currentBets": Array [
    0.5,
  ],
  "dealerPosition": 1,
  "id": "200000000000000000000001",
  "isDone": false,
  "isStarted": true,
  "phases": Array [
    Object {
      "cards": Array [],
      "moves": Array [],
      "type": "pre-flop",
    },
  ],
  "players": Array [
    Object {
      "hand": Array [
        Object {
          "card": 2,
          "color": "c",
        },
        Object {
          "card": 2,
          "color": "s",
        },
      ],
      "money": 50,
      "seat": 1,
      "user": Object {
        "id": "000000000000000000000001",
        "username": "user1",
      },
    },
    Object {
      "hand": null,
      "money": 50,
      "seat": 2,
      "user": Object {
        "id": "000000000000000000000002",
        "username": "user2",
      },
    },
    Object {
      "hand": null,
      "money": 50,
      "seat": 3,
      "user": Object {
        "id": "000000000000000000000003",
        "username": "user3",
      },
    },
  ],
  "pot": 0,
  "stakes": 50,
  "tableId": "100000000000000000000001",
}
`;

exports[`should return a game with no players 1`] = `
Object {
  "betMap": Object {},
  "currentBets": Array [],
  "dealerPosition": -1,
  "id": "200000000000000000000001",
  "isDone": false,
  "isStarted": false,
  "phases": Array [
    Object {
      "cards": Array [],
      "moves": Array [],
      "type": "pre-flop",
    },
  ],
  "players": Array [],
  "pot": 0,
  "stakes": 0,
  "tableId": "100000000000000000000001",
}
`;
